### ISSUE TITLE
Right-click run for tests not working for colocated tests

### ISSUE THREAD

**Issue by rorystephenson** [flutter/flutter-intellij#6046]
I'm migrating my tests to be colocated with the code they are testing as I prefer this approach.

I've noticed that right-click will run the tests with `flutter test` when they are in my test/ directory but not when they are colocated with the code. My colocated tests are in a test/ subfolder of the directory where the code they are tested is located. So tests for lib/src/foo/bar.dart are in lib/src/foo/test/bar_test.dart

## Steps to Reproduce

1. Create an identical test file in `test/foo/bar_test.dart` and `lib/src/foo/test/bar_test.dart`.
2. Right-click the file in `test/foo/bar_test.dart` and run it, it runs the tests with `flutter test` as expected.
3. Right-click the file in `lib/src/foo/test/bar_test.dart` and run it, it runs the tests with `dart test` when it should be `flutter test`. ~~Note that when clicking the gutter icon inside the file to run an individual test it correctly uses `flutter test`~~ I just retried this and it's not working, I think I was mistaken.

### example test file

```dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('right click run test', () {
    test('dummy test', () {
      expect(1, equals(1));
    });
  });
}
```

## Version info

<details>
<summary>flutter doctor</summary>

```
[✓] Flutter (Channel stable, 2.10.3, on macOS 11.5.2 20G95 darwin-x64, locale en-GB)
    • Flutter version 2.10.3 at /Users/rorystephenson/sdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (2 weeks ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/rorystephenson/sdk/android/
    • Platform android-32, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/rorystephenson/sdk/android
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] VS Code (version 1.65.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.34.0

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 99.0.4844.74

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by rorystephenson** [comment]
I've just had a quick look at the source code and I think [this code is to blame](https://github.com/flutter/flutter-intellij/blob/6350cf8d1da5fc8d4739a2519ca22dfdc037bfcb/flutter-idea/src/io/flutter/FlutterUtils.java#L194-L204). It check if the file is in one of the following top-level directories:

 - `test/`
 - `integration_test/`
 
This is out-of-sync with the Dart linter's requirement for a file to be considered a test file which is that the path of the file should [**contain**](https://github.com/dart-lang/sdk/blob/9c0689a887aedb5313131cd6a0fefeb9b2d2dbe8/pkg/analyzer/lib/src/lint/linter.dart#L396) one of the following [directories](https://github.com/dart-lang/sdk/blob/9c0689a887aedb5313131cd6a0fefeb9b2d2dbe8/pkg/analyzer/lib/src/lint/linter.dart#L282-L288):

 - `test/`
 - `integration_test/`
 - `test_driver/`
 - `testing/`
 
 This means that the analyzer considers for example `lib/src/foo/test_driver/bar_test.dart` to be a test file whilst flutter-intellij does not.
 
 Making flutter-intellij's detection for test files match the Dart analyzer's would improve consistency and allow colocated test files to be run via right-click.

**Comment by stevemessick** [comment]
Thanks for the detailed report!

**Comment by stevemessick** [comment]
Actually, this feature should already be working. It was implemented in #5770. Set the Flutter plugin option to recognize files ending in "_test.dart" as tests, and mark the "test" directories as a Test Sources Root. You may need to delete any auto-generated run configs. There may be a bug in the implementation, because I thought it was supposed to work in "lib" without having to mark anything as a Test Sources Root (since "lib" is already marked as a Sources Root). That would be a separate issue, though.

Try that and let us know if it helps.

**Comment by rorystephenson** [comment]
Thanks @stevemessick that seems to work! Ideally I wouldn't have to mark all of my test directories as test source routes but that's a minor bother.
